Crop Rework 2.0

C0bra5 opened the pull request:
This rework isn't a continuation of the old rework started 2 years ago, but a new effort from scratch.

The rework is now somewhat playable. You should be able to just slot it into the daily experimental build of the pack to start testing things!

More info can be found here:
- Design Doc: https://docs.google.com/document/d/10a0WaCF1O98jkuO_dKjxbxAeiJkMuQiBGz7h7j5hNec/edit?usp=sharing
- Data Sheets: https://docs.google.com/spreadsheets/d/10FevOcA-KWmxdp0OU7qPAgXp-a5qhE2JOjhG1kvndBo/edit?gid=200076011#gid=200076011

Progress Tracker:
- [X] Waila Compat
- [X] Seed Item Generation
- [x] All crops re-added
  - [x] Vanilla Crops
  - [x] Vanilla Flowers
  - [x] IC2 Crops
  - [x] Crops++ Crops
    - [x] GregTech Crops (Gardinia + Stone Lilies)
    - [x] BoP Crops
    - [x] Crops++ Crops
    - [x] Thaum Crops
    - [x] Natura Crops
    - [x] TiC Crops
    - [x] Twilight Forest Crops
    - [x] Witchery Crops
  - [x] GregTech Crops (The ones actually from GregTech)
  - [x] Good Generator Crops (Salty Root)
  - [x] GT++ Crops (hemp) 
- [X] Import and fix existing IC2 crop art
- [X] Deterministic Breeding
- [X] Agricraft Style Spreading
- [X] Add aspects to items
- [x] NEI Compat 
  - [X] Crop List Tab
  - [X] Deterministic Breeding Tab
  - [X] Mutation Pool Tab
  - [X] Data Dumping Tools
  - [x] Duplicator Tab
  - [x] Breeder Tab
  - [x] Gene Extractor Tab
  - [x] Synthesizer Tab
- [X] Create Breeding Tree
- [x] Crop Manager Updated 
- [x] Duplicator singleblock
- [x] Breeder singleblock
- [x] Gene Extractor Singleblock
- [x] Synthesizer singleblock
- [x] Update/Add Recipes
  - [x] Fertilizer Recipes
  - [x] Material Leaf Recipes
  - [x] New Liquid Fertilizer Recipes
  - [x] Single Block Recipes
  - [x] Multiblock Components Recipes
  - [x] Environmental Module Recipes
- [x] Textures
  - [x] Redo all crop textures that use anything from IC2
    - All textures have been redone, not applied yet
  - [x] Seed Generator
  - [x] Crop Breeder
  - [x] Crop Gene Analyzer
  - [x] Crop Synthesizer
  - [x] Seed Bed
  - [x] New Casing
  - [x] Environmental Enhancement Unit
  - [x] Growth Acceleration Unit
  - [x] Overclocked Growth Acceleration Unit
  - [x] Fertilizer Unit
  - [x] Advanced Harvesting Unit
  - [x] Stone Mug + Filled Mugs
    - All textures have been redone, not applied yet
- [x] Rename all IC2 crops with original names.
- [x] Implement trampling mechanic.
- [x] Implement crop disease mechanic.
- [x] Implement disease-curing mechanic.
- [x] Implement TiC Scythe Harvesting Compat
- [x] Finalize growth formula/mechanic.
- [x] New crop multi
  - [x] Upgrades Implemented
    - [x] Environmental Enhancement Unit
    - [x] Growth Acceleration Unit
    - [x] Fertilizer Unit
    - [x] Advanced Harvesting Unit
    - [x] Overclocked Growth Acceleration Unit
  - [x] Structure Code Implemented
  - [x] Processing Logic Implemented
  - [x] Texture Pass
- [x] Re-implement IC2 Stone Mug + Filled Versions
- [x] World Migrations
  - [x] Postea PR merged: https://github.com/GTNewHorizons/Postea/pull/18
    - [x] IC2 Migrations Complete
    - [x] GT5u Migrations Complete
    - [x] Crops++ Migrations Complete
  - Migrations are currently implemented without the refactor PR, will need to update once merged
- [x] Scanner Recipe Integration
  - [x] Scanner Refactor PR merged: https://github.com/GTNewHorizons/GT5-Unofficial/pull/5677 
  - [x] Scanner Handler Added
- [x] Drop rate testing


Integration PRs
- GT5u: https://github.com/GTNewHorizons/GT5-Unofficial/pull/5652
- NewHorizonCoreMod: https://github.com/GTNewHorizons/NewHorizonsCoreMod/pull/1604
- Binnie's mod: https://github.com/GTNewHorizons/Binnie/pull/82
  - Change coffee reference to GT's ore dict
- Cooking for Blockheads: https://github.com/GTNewHorizons/CookingForBlockheads/pull/59
  - Update Coffee references to GT5u's ore dict
- Raicraft: pending
  - Need to remove/disable IC2 crop support from it.
- GT-New-Horizons-Modpack: https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/pull/23254
  - Add New Quest book tab
  - Remove old Crop tab
  - Remove the following mods from the pack:
    - [Crops-plus-plus](https://github.com/GTNewHorizons/Crops-plus-plus)
    - [CropLoadCore](https://github.com/GTNewHorizons/CropLoadCore)
    - [IC2 Crop-Breeding Plugin](https://www.curseforge.com/minecraft/mc-mods/ic2-nei-crop-plugin)
  - Add CropsNH to pack
- Witching Gadgets: https://github.com/GTNewHorizons/WitchingGadgets/pull/120
  - Uses the IC2 coffee item list entry

C0bra5 committed:
add vanilla crops

C0bra5 committed:
trim unused deps

C0bra5 committed:
remove unused textures

C0bra5 committed:
crop transfer fixes

C0bra5 committed:
ged rid of unused crop card constructor

C0bra5 committed:
add vanilla flowers

C0bra5 committed:
add ic2 crops

C0bra5 committed:
Add GT from Crops++

C0bra5 committed:
implement crop collision handling and collision damage

C0bra5 committed:
add all ore berries

C0bra5 committed:
add witchery crops

C0bra5 committed:
spotless apply

C0bra5 committed:
add bop crops

C0bra5 committed:
spotless apply

C0bra5 committed:
add bop glowshrooms and compound soil list

C0bra5 committed:
add natura dep

C0bra5 committed:
un comment commit deps

tired of having to double check this file on every commit

C0bra5 committed:
somthing about capitalisation

C0bra5 committed:
Update CropKnightmetalBerries.java

C0bra5 committed:
prevent re 31 crops from turning into weeds

C0bra5 committed:
add natura crops

C0bra5 committed:
fix magic seed shape base

C0bra5 commented:
@Lyfts This PR is not ready in any way, and is still _very much_ under progress.

Lyfts commented:
I honestly have no clue how I ended up undrafting lol, just stopped by for a few secs earlier to see what was changed

Ethryan commented:
11k lines and counting…

This will be “fun” to review…

C0bra5 committed:
trim unused configs and make crop tile use them

C0bra5 committed:
more config cleanup

C0bra5 committed:
On this day, we welcome his screaminess back into the realm of the cube

C0bra5 committed:
hodgepodge and angelica compat

it's still broken because both hodgepodge and angelica are broken to begin with

C0bra5 committed:
make essence berries animate again

C0bra5 committed:
add config to disable the stepped on cries

C0bra5 committed:
remaining core crops++ crops

C0bra5 committed:
add the last crops++ crops

C0bra5 committed:
prep for gt crop import

C0bra5 committed:
angelica compat

The fish must blub,
Big thanks to cookie brigade for fixing angelica.

C0bra5 committed:
port hemp and salty root

C0bra5 committed:
all gregtech crops added

testing of crops to be done tommorow

C0bra5 committed:
nei and creative tab fix

C0bra5 committed:
misc fixes

C0bra5 committed:
fixes pt2

C0bra5 committed:
breeding system

C0bra5 committed:
add dumpers

C0bra5 committed:
better names for the dump files

C0bra5 committed:
add starwart material leaf, clarify loader function names

C0bra5 committed:
nei crop list check

C0bra5 committed:
Stop NEI from acting as a budget scanner

C0bra5 committed:
update milkwart sprite to spore shape

C0bra5 committed:
make starwart drop new material leaf

C0bra5 committed:
move starwart skull drop to witheread

C0bra5 committed:
add note for wither skull drop for future consideration

C0bra5 committed:
add rubber wood sappling

C0bra5 committed:
deterministic breeding tree

C0bra5 committed:
mutation rebalancing

C0bra5 committed:
reuse old attributes for mutation pool

C0bra5 committed:
spotless apply

C0bra5 committed:
NEI breeding compat

C0bra5 committed:
crop card housekeeping

C0bra5 committed:
cleanup

C0bra5 committed:
add simulate to crop sticky apply functions

C0bra5 committed:
crop manager v2 added

C0bra5 committed:
spotless apply

C0bra5 committed:
add fertilizer fluid and override ic2 fert item in recipes pt1

C0bra5 committed:
update build script add core mod runtime dep for testing

C0bra5 committed:
recipe transfer checkpoint

C0bra5 committed:
dep update, crop manager fix + upgrade.

C0bra5 committed:
recipe jail continues

C0bra5 committed:
spotless apply

C0bra5 committed:
fix buildscript

C0bra5 committed:
add item aspects

C0bra5 committed:
remove argentia from reed pool

how did that get in there?

C0bra5 committed:
update buildscripts, update code to match dependency changes

C0bra5 committed:
add circuit option to get impure dust from material leaves

C0bra5 committed:
recipe fixes

C0bra5 committed:
add seed replicator single block

- removed uxv crop manager since uxv machines don't really exist
- renamed crop manager item names to be more in line with everything else

C0bra5 committed:
enforce no-cross and no breed thresholds

C0bra5 committed:
add duplication catalysts for existing material crops

C0bra5 committed:
fix meta map

C0bra5 committed:
fix broken permutations in crop produce nei tab

C0bra5 committed:
add seed generator to nei

C0bra5 committed:
update bop dep

C0bra5 committed:
update buildscript

C0bra5 committed:
cleanup compat

C0bra5 committed:
cleanup unused utils and item renderers

C0bra5 committed:
Fix duplicated mutation parent and add verification.

C0bra5 committed:
fix incorrect nbt reading for gain

C0bra5 committed:
add crop breeder single block and NEI tab

C0bra5 committed:
add crop breeder recipe

C0bra5 committed:
spotless apply

C0bra5 committed:
Make breeding machine recipe tier a crop component

C0bra5 committed:
Fix Enderbloom growth requirements

- Added endstone block under
- Changed duplication catalyst from ender pearl to ender eye to nudge towards the HV semi-tier gate a bit more.

C0bra5 committed:
remove call to marked for removal function.

C0bra5 committed:
Add Crop Gene Extractor

C0bra5 committed:
refactor frontend and backend packages

C0bra5 committed:
simplify lang ids

C0bra5 committed:
Add Seed Synthesizer

C0bra5 committed:
get renderer fixes

C0bra5 committed:
make the item list IItemContainer compatible

C0bra5 committed:
Organise TE files before multi implementation begins

C0bra5 committed:
fix broken resources

C0bra5 committed:
Add Structure for Industrial Farm

C0bra5 committed:
spotless apply

C0bra5 committed:
rebalances, bug fixes, cleanups, multi-prep

C0bra5 committed:
better tooltips and fix seed capacity calculations

C0bra5 committed:
spotless apply

C0bra5 committed:
add environmental modules, improve IF GUI

C0bra5 committed:
update deps to latest for MUI2 stuff

C0bra5 committed:
remove unnecessary mod buttons from IF UI.

C0bra5 committed:
add structure lib runtime dep.

I need to check if the preview works right in NEI

C0bra5 committed:
add growth mechanics

C0bra5 committed:
implement farm mode processing logic

C0bra5 committed:
Remove lang files that aren't being worked on.

I probably could have kept the French one alive, but I'd rather just retranslate everything once it's done.

C0bra5 committed:
cleanup mod id references

C0bra5 committed:
standardize mod id references

C0bra5 committed:
spotless apply

C0bra5 committed:
bug fixing, remove unused code from agricraft

C0bra5 committed:
apply nutrient scale in growth speed function

C0bra5 committed:
growth formula bug fixes

C0bra5 committed:
add scanner descriptions

C0bra5 committed:
scan results, machine only reqs, refactors and more QA stuff

Instead of preventing crops from growing outside of the machine outright, I've opted to prevent them from being harvested instead. This should help with stating machine only crops outside of the machine.

C0bra5 committed:
dumper improvements

C0bra5 committed:
add liked biomes to crops

C0bra5 committed:
more dumper fixes

C0bra5 committed:
Texture updates for IC2 textures.

Courtesy of coolprez on the GTNH discord

C0bra5 committed:
update hemp texture

C0bra5 committed:
forgot the first textrure

C0bra5 committed:
fix missing drops count at low gain

C0bra5 committed:
typo fixes

C0bra5 committed:
add recipes for the multi

C0bra5 committed:
fix micadia material leaf name

C0bra5 committed:
The great naming refactor + 2 more stone lilies

removed ic2 crop names
reorganized a couple things
added deepslate and tuff stone lilies

C0bra5 committed:
cleanup unused crop textures from agricraft

C0bra5 committed:
removed unused seed textures from agricraft

C0bra5 committed:
remove all unused agricraft textures

C0bra5 committed:
unyeet the cropstick texture

C0bra5 committed:
fix deepslate lily growth req

C0bra5 committed:
add trampling mechanic + spotless apply

C0bra5 committed:
TiC Scythe Compat

C0bra5 committed:
fix startwart

C0bra5 committed:
add plant cure

C0bra5 committed:
remove water module

C0bra5 committed:
add recipes for environmental modules

C0bra5 committed:
fix invalid drop count on ardite berry

C0bra5 committed:
fix broken textures

C0bra5 committed:
fix load order for ore dicts and crops

C0bra5 committed:
fix broken milk recipe

C0bra5 committed:
fix EIG compat

C0bra5 committed:
remove debugger

C0bra5 committed:
Remove NAN Certificate

Moving to Core Mod

C0bra5 committed:
rename magnifying glass to plant lens

C0bra5 committed:
spotless apply + build fix

C0bra5 committed:
reduce slimy bonsai drop to 4 logs per drop

C0bra5 committed:
add weed-ex can and poison powder + some fertilizer recipe fixes

C0bra5 committed:
build fix

C0bra5 committed:
add textures for weedex + poison powder

C0bra5 committed:
make sb machines not use ic2 noises

C0bra5 committed:
lessen reliance on some GT APIs

C0bra5 committed:
bug fixes

C0bra5 committed:
add reinforced spade

C0bra5 committed:
spade feature parity

C0bra5 committed:
fix scythe weed clearing

C0bra5 committed:
doc

C0bra5 committed:
Migrate liquid fert to CropsNH

It doesn't have a use right now so I'm fine with nuking what ever used to contain it.

C0bra5 committed:
final gt5u fertilizer recipes migration, automatic migrations, bug fixes

C0bra5 committed:
fix bad config values

C0bra5 committed:
sicken crops instead of replacing them with weeds.

C0bra5 committed:
spotless apply

C0bra5 committed:
Merge remote-tracking branch 'upstream/master'

C0bra5 committed:
remove ccc dep

C0bra5 committed:
name standardizations

C0bra5 committed:
fix bad config

C0bra5 committed:
yeet debug code

C0bra5 committed:
hide weed and migrator crop in nei

C0bra5 committed:
fix glowshroom soil type

C0bra5 committed:
make global soil list available for weeds and migrator

C0bra5 committed:
fix coffe drop?

C0bra5 committed:
add vine as alternative seed

C0bra5 committed:
add soil type to crop dump

C0bra5 committed:
Make bamboo use the same soil type as bonsai crops.

C0bra5 committed:
spotless apply

C0bra5 committed:
move ore dict registration to pre-init

C0bra5 committed:
ensure pams loads before cropsnh

C0bra5 committed:
Add support for ExtraUtils watering can compat for watering

Dirty implementation, but I don't have an event to listen to and it's getting replaced with UiE so a mixin feels superfluous for something that's going away soon-ish anyway.

C0bra5 committed:
Beginnings of UiE Compat

the crop stick logic will get removed once the watering can event is implemented correctly.

C0bra5 committed:
add missing alternate seeds

C0bra5 committed:
migrate terra wart recipes from gt5u

C0bra5 committed:
add migration for gtpp crop manager

C0bra5 committed:
add machine type to single block machine tooltips

C0bra5 committed:
migrate more stuff over from gt5u and core mod

C0bra5 committed:
fix crash/typo

C0bra5 committed:
spotless apply

C0bra5 committed:
server crash fix

C0bra5 committed:
spotless apply

C0bra5 committed:
add sugar beet mortar recipe

C0bra5 committed:
update weedex reference to crops nh's

C0bra5 committed:
add seed scanner fake recipe

once the scanner refactor pr is in, I'll add the handler

C0bra5 committed:
update coffee references

C0bra5 committed:
standardize all mod id references to lower case

C0bra5 committed:
call em by what people call em

C0bra5 committed:
add ore dict for tea leaf

C0bra5 committed:
add breeding for slimy bonsai

C0bra5 committed:
fix broken output when scanning empty crop sticks

C0bra5 committed:
migrate ore dict defs from crop load core

C0bra5 committed:
remove tea leaf material leaf

C0bra5 committed:
fix broken power consumption from crop manager

C0bra5 committed:
fix localizations, add max tomato

C0bra5 committed:
buff seed bed capacity

C0bra5 committed:
for loc fixes

C0bra5 committed:
max tomato image

C0bra5 committed:
add listAllSeed to all CropsNH seeds

C0bra5 committed:
fix max tomato saturation

C0bra5 committed:
make mica crop accept cupronickel coils as underblock

C0bra5 committed:
allow all space rocks for space flower underblock

- The GC cheese ore will no longer count as a valid block.

C0bra5 committed:
fix potential crash caused by bad replication catalyst

better be failsafe than fail hard

C0bra5 committed:
fix broken localization

C0bra5 committed:
internalize getModItem, reduce dependencies, remove removed recipes

- remove recipes migrated from other mods that no longer exist due to unused content removals
- fix slimy sappling drops to use correct block/item
- internalize getModItem
- use oredicts when ever applicable
- update dependencies

C0bra5 committed:
set/map optimisations

C0bra5 committed:
fix incorrect chat formatting

C0bra5 committed:
fix incorrect item id

C0bra5 committed:
limit larger CR recipes to LCR

C0bra5 committed:
Migrate an optimize IC2 forestry multi-farm compat.

removes the ability to stack farms horizontally

C0bra5 committed:
increase fertilizer cost of crop forestry multifarms

C0bra5 committed:
cleanup

C0bra5 committed:
fix server crash

C0bra5 committed:
use better maps/sets

C0bra5 committed:
allow mutation pools to work with 2 of the same crop

C0bra5 committed:
spotless apply

C0bra5 committed:
migrate coffee from IC2

texture by June from the GTNH discord

C0bra5 committed:
spotless apply

C0bra5 committed:
shrink textures

C0bra5 committed:
fix typo

C0bra5 committed:
Enforce growth requirement planting breeding offspring.

C0bra5 committed:
fix incorrect weed-ex usage

C0bra5 committed:
spotless apply

C0bra5 committed:
fix crash when planting stuff

C0bra5 committed:
fix stat variation when parents have fertilizer

C0bra5 committed:
increase code readability

C0bra5 committed:
code readability

C0bra5 committed:
fix spade breaking speed

C0bra5 committed:
fix seed drop with immature crop and spade.

C0bra5 committed:
fix manual fert application threshold.

C0bra5 committed:
change item fertilisation logic to be a bit fairer

C0bra5 committed:
fix deepslate lily mutation

C0bra5 committed:
add mutation pools to snowbell and balzereed

C0bra5 committed:
Prevent immature crops from transferring stats.

C0bra5 committed:
fix incorrect parent size calculation for breeder machine and mutations

C0bra5 committed:
Update l10n to to fix crash on latest daily

C0bra5 committed:
fix incorrect aspects for crop sticks

should be an auto-generated value, based on the crafting recipe, will nerf the amount of arbor you get since the recipe cost is going down 6 -> 1

C0bra5 committed:
Add single block scanner handler

C0bra5 committed:
update migrations to work with the postea refactor PR

- this also removes all dev migrations so anyone who used a version prior to this commit might get some odd stuff if they didn't run the build from just before this commit, i think daily 395 should be compatible with it.

C0bra5 committed:
make salty root easier to see in default UI

C0bra5 committed:
postea handler fix and dep update

C0bra5 committed:
Add new crop textures

Made by june_username on the discord!

boubou19 reviewed (approved):
:shipit:

C0bra5 committed:
Dissable runtime deps so the CI passes

ok by the boublerone
